File: api-code-examples/nodejs/docs-set.js
import assert from 'node:assert'
import { Iroh } from '@number0/iroh'

// Create a node
const node = await Iroh.memory()
const nodeId = await node.net.nodeId()
console.log(`Started iroh node: ${nodeId}`)

// Create a document
const doc = await node.docs.create()
console.log(`Created doc: ${doc.id()}`)

// Use the default author to write
const author = await node.authors.default();

// set a value in the document
const key = Array.from(Buffer.from("key"))
const value = Array.from(Buffer.from("value!"))
await doc.setBytes(author, key, value)

// get a key
const entry = await doc.getExact(author, key, false)

// read the value back
const res = await node.blobs.readToBytes(entry.hash)

// returned value matches the original
assert.deepEqual(value, res)

console.log(`Got value: ${Buffer.from(res).toString()}`)

File: api-code-examples/rust/docs-set.rs
use anyhow::{anyhow, Result};

use iroh::docs::store::Query;

#[tokio::main]
async fn main() -> Result<()> {
    let node = iroh::node::Node::memory().enable_docs().spawn().await?;

    // create a document
    let author = node.authors().default().await?;
    let doc = node.docs().create().await?;

    // set the key "key" to "value"
    let key = b"key";
    let value = b"value";
    doc.set_bytes(author, key.to_vec(), value.to_vec()).await?;
    println!("key is set!");

    // read the value back
    let query = Query::key_exact(key);
    let entry = doc
        .get_one(query)
        .await?
        .ok_or_else(|| anyhow!("entry not found"))?;
    let content = entry.content_bytes(node.client()).await?;

    println!("value bytes: {:?}", content);

    Ok(())
}
